Define types for new openssh executables

zpytela opened the pull request:
/usr/libexec/openssh/sshd-auth and /usr/libexec/openssh/sshd-session

zpytela committed:
Define types for new openssh executables

In openssh-9.9 (Fedora 41), the sshd server has been split into a listener
binary, sshd(8), and a per-session binary, "sshd-session". This allows for
a much smaller listener binary, as it no longer needs to support the SSH
protocol. As part of this work, support for disabling privilege separation
(which previously required code changes to disable) and disabling
re-execution of sshd(8) has been removed.

In openssh-10.0 (Fedora 43), the code responsible for the user
authentication phase of the protocol has been moved from the
per-connection sshd-session binary to a new sshd-auth binary.
Splitting this code into a separate binary ensures that the crucial
pre-authentication attack surface has an entirely disjoint address space
from the code used for the rest of the connection. It also yields a small
runtime memory saving as the authentication code will be unloaded after
the authentication phase completes. This change should be largely invisible
to users, though some log messages may now come from sshd-auth instead
of sshd-session.

sshd reads host secret keys and execs sshd-session on incoming connections.
After successful auth (happening in sshd-auth if it exists), sshd-session
forks again and drops privileges.
sshd-session processes pam auth.
sshd, sshd-session require access to ssh-agent auth socket.

/usr/libexec/openssh/sshd-auth
/usr/libexec/openssh/sshd-session

bachradsusi commented on policy/modules/roles/unconfineduser.te:
```diff
@@ -116,6 +116,7 @@ tunable_policy(`selinuxuser_execmod',`
 tunable_policy(`unconfined_login',`
 	corecmd_shell_domtrans(unconfined_login_domain,unconfined_t)
 	ssh_dyntransition_to(unconfined_t)
```
Given that the transition is done by `ssh-session` now, would it make sense to drop this?

zpytela replied:
It would, as well as some more cleanup for the sshd_t and sshd_net_t domains.

bachradsusi commented on policy/modules/services/ssh.if:
```diff
@@ -310,6 +310,30 @@ template(`ssh_server_template',`
+	')
+
+	typeattribute $1 ssh_session_dyntransition_domain;
+	#domain_type($1)
```
Is it a leftover comment?

zpytela replied:
Yes, it is, there might be more like this given this PR is still a draft.

bachradsusi commented on policy/modules/services/ssh.te:
```diff
@@ -66,6 +68,56 @@ mls_trusted_object(sshd_t)
+#        ssh_session_dyntransition_to(unconfined_t)
+#')
+
+type sshd_auth_t;
```
I believe that `sshd_net_t` is basically `sshd_auth_t`

The current Fedora openssh version just changes context https://src.fedoraproject.org/rpms/openssh/blob/rawhide/f/0002-openssh-6.6p1-privsep-selinux.patch - imho this patch is based on the previous version without `sshd-auth` which should not be necessary anymore when `sshd-auth` is correctly confined.

But I could be missing something important.

zpytela replied:
I see only dynamic transitions to sshd_net_t from session and auth, while session->auth is a type transition on a file execution.

vmojzis commented:
Rebased to v42.8

vmojzis committed:
Add default contexts for sshd-seesion

In openssh-9.9 (Fedora 41), the sshd server has been split into a listener
binary, sshd(8), and a per-session binary, "sshd-session".

sshd-session runs in a new domain sshd_session_t, which is now used to
transition into an appropriate login context.

Signed-off-by: Vit Mojzis <vmojzis@redhat.com>

vmojzis commented:
Seems to work properly after the last update (at least in targeted):
```
ssh staff@10.0.186.197
staff@10.0.186.197's password: 
*** 1minutetip system created by vmojzis - Mon Sep 15 03:57:38 PM EDT 2025 ***
Last login: Mon Sep 15 16:12:29 2025 from 10.45.224.92
[staff@vm-10-0-186-197 ~]$ id 
uid=1001(staff) gid=1001(staff) groups=1001(staff) context=staff_u:staff_r:staff_t:s0-s0:c0.c1023
[staff@vm-10-0-186-197 ~]$ exit
logout
Connection to 10.0.186.197 closed.
╭─vmojzis@fedora ~/devel/github/fedora-selinux/selinux-policy  ‹confine-new-sshd*› 
╰─$ ssh sysadm@10.0.186.197
sysadm@10.0.186.197's password: 
*** 1minutetip system created by vmojzis - Mon Sep 15 03:57:38 PM EDT 2025 ***
[sysadm@vm-10-0-186-197 ~]$ id
uid=1002(sysadm) gid=1002(sysadm) groups=1002(sysadm) context=sysadm_u:sysadm_r:sysadm_t:s0-s0:c0.c1023
[sysadm@vm-10-0-186-197 ~]$ exit
logout
Connection to 10.0.186.197 closed.
╭─vmojzis@fedora ~/devel/github/fedora-selinux/selinux-policy  ‹confine-new-sshd*› 
╰─$ ssh user@10.0.186.197  
user@10.0.186.197's password: 
*** 1minutetip system created by vmojzis - Mon Sep 15 03:57:38 PM EDT 2025 ***
[user@vm-10-0-186-197 ~]$ id
uid=1003(user) gid=1003(user) groups=1003(user) context=user_u:user_r:user_t:s0
[user@vm-10-0-186-197 ~]$ ps -axZ | grep sshd
system_u:system_r:sshd_t:s0-s0:c0.c1023 1688 ?   Ss     0:00 sshd-session: root [priv]
unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 1708 ? S   0:00 sshd-session: root@pts/0
system_u:system_r:sshd_t:s0-s0:c0.c1023 4518 ?   Ss     0:00 sshd: /usr/sbin/sshd -D [listener] 0 of 10-100 startups
system_u:system_r:sshd_session_t:s0-s0:c0.c1023 5562 ? Ss   0:00 sshd-session: user [priv]
user_u:user_r:user_t:s0            5579 ?        S      0:00 sshd-session: user@pts/1
user_u:user_r:user_t:s0            5646 pts/1    S+     0:00 grep --color=auto sshd
[user@vm-10-0-186-197 ~]$ 

ausearch -m AVC -ts recent
----
time->Mon Sep 15 16:55:25 2025
type=PROCTITLE msg=audit(1757969725.271:809): proctitle=737368642D73657373696F6E3A2073797361646D205B707269765D
type=SYSCALL msg=audit(1757969725.271:809): arch=c000003e syscall=1 success=yes exit=42 a0=a a1=55ee29298590 a2=2a a3=0 items=0 ppid=5494 pid=5511 auid=1002 uid=1002 gid=1002 euid=1002 suid=1002 fsuid=1002 egid=1002 sgid=1002 fsgid=1002 tty=(none) ses=12 comm="sshd-session" exe="/usr/libexec/openssh/sshd-session" subj=sysadm_u:sysadm_r:sysadm_t:s0-s0:c0.c1023 key=(null)
type=AVC msg=audit(1757969725.271:809): avc:  denied  { dyntransition } for  pid=5511 comm="sshd-session" scontext=system_u:system_r:sshd_session_t:s0-s0:c0.c1023 tcontext=sysadm_u:sysadm_r:sysadm_t:s0-s0:c0.c1023 tclass=process permissive=1

```
The AVC is just for sysadm_t and can be allowed using a boolean.

zpytela commented:
Merging, thanks for the collaboration.

AkihiroSuda commented on policy/modules/services/ssh.te:
```diff
@@ -66,6 +68,116 @@ mls_trusted_object(sshd_t)
+term_setattr_all_ttys(sshd_session_t)
+term_relabelto_all_ptys(sshd_session_t)
+
+type sshd_auth_t;
```
Doesn't seem to work with vsock:
- https://github.com/fedora-selinux/selinux-policy/issues/2955

```console
# ausearch -m AVC,USER_AVC -ts recent
----
time->Fri Nov 21 04:10:25 2025
type=AVC msg=audit(1763698225.743:248): avc:  denied  { getattr } for  pid=1355 comm="sshd-auth" scontext=system_u:system_r:sshd_auth_t:s0-s0:c0.c1023 tcontext=system_u:system_r:sshd_t:s0-s0:c0.c1023 tclass=vsock_socket permissive=1
```
